README.md
Addy: a modern library for working with email addresses
=======================================================
[![CI](https://github.com/pjones/addy/workflows/CI/badge.svg)](https://github.com/pjones/addy/actions)
[![GitHub tag (latest by date)](https://img.shields.io/github/v/tag/pjones/addy?label=release)](https://github.com/pjones/addy/releases)
[![Hackage](https://img.shields.io/hackage/v/addy?logo=haskell)](https://hackage.haskell.org/package/addy)
[![GitHub](https://img.shields.io/github/license/pjones/addy)](https://github.com/pjones/addy/blob/master/LICENSE)

A full-featured library for parsing, validating, and rendering email
addresses.

Decoding
--------

```haskell
Addy.decode "example@example.com"
-- Right (EmailAddr "example@example.com")

Addy.decode "我買@屋企.香港"
-- Right (EmailAddr "\25105\36023@\23627\20225.\39321\28207")

Addy.decode "Mary Smith <mary@example.net> (hi there!)"
-- Right (EmailAddr "Mary Smith <mary@example.net> (hi there!)")

Addy.decode "example@[127.0.0.1]"
-- Right (EmailAddr "example@[127.0.0.1]")
```

Encoding
--------

```haskell
Addy.encode address
-- "example@example.com"

Addy.decode "Mary Smith <mary@example.net> (hi there!)"
  & second Addy.encodeFull
-- Right "Mary Smith <mary@example.net> (hi there!)"
```


CHANGES.md
# Release History

## 0.1 (May 8, 2020)

  * Initial release

  * Minor updates:

    - 0.1.0.1 (October 29, 2020): Update dependencies


test/Main.hs
-- |
--
-- Copyright:
--   This file is part of the package addy. It is subject to the license
--   terms in the LICENSE file found in the top-level directory of this
--   distribution and at:
--
--     https://code.devalot.com/open/addy
--
--   No part of this package, including this file, may be copied,
--   modified, propagated, or distributed except according to the terms
--   contained in the LICENSE file.
--
-- License: BSD-2-Clause
module Main
  ( main,
  )
where

import qualified ParserTest
import qualified RenderTest
import Test.Tasty

-- | Main.
main :: IO ()
main =
  defaultMain $
    testGroup
      "Email"
      [ ParserTest.test,
        RenderTest.test
      ]


test/TestData.hs
-- |
--
-- Copyright:
--   This file is part of the package addy. It is subject to the license
--   terms in the LICENSE file found in the top-level directory of this
--   distribution and at:
--
--     https://code.devalot.com/open/addy
--
--   No part of this package, including this file, may be copied,
--   modified, propagated, or distributed except according to the terms
--   contained in the LICENSE file.
--
-- License: BSD-2-Clause
module TestData
  ( IsEmailTest (..),
    IsEmailCat (..),
    isEmailTests,
    rfc5322Examples,
    rfc5322ObsExamples,
    wikipediaIntExamples,
    miscExamples,
    genShortEmail,
    genLongEmail,
  )
where

import Addy.Internal.Char (utf8NonAscii)
import Data.Aeson (FromJSON)
import qualified Data.Aeson as Aeson
import Data.Char
import qualified Data.Text as Text
import Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
import Network.HTTP.Types.URI (urlDecode)
import Test.Tasty.HUnit

data IsEmailTest = IET
  { ietId :: Int,
    ietAddr :: Text,
    ietCat :: Text,
    ietDia :: Text
  }
  deriving stock (Generic, Show)
  deriving anyclass (FromJSON)

data IsEmailCat
  = CatError
  | CatDeprec
  | CatOkay
  deriving (Eq, Show)

isEmailTests :: IO [(IsEmailTest, IsEmailCat)]
isEmailTests = do
  let fileName = "test/isemail.json"
  tests <- (Aeson.decode <$> readFileLBS fileName) >>= \case
    Nothing -> assertFailure "failed to load the isemail.json file"
    Just ts -> pure ts
  pure $ map (\t -> (decodeAddr t, decodeCat t)) tests

-- | The XSL file URL-encodes the email addresses.  Additionally, the
-- XML file encodes some characters by adding 0x2400 to them.
decodeAddr :: IsEmailTest -> IsEmailTest
decodeAddr iet@IET {ietAddr} =
  iet
    { ietAddr = unCtrl (decode ietAddr)
    }
  where
    decode :: Text -> Text
    decode = decodeUtf8 . urlDecode False . encodeUtf8
    unCtrl :: Text -> Text
    unCtrl =
      let f c =
            if c >= '\x2400' && c <= '\x241f'
              then chr (ord c - 0x2400)
              else c
       in Text.map f

decodeCat :: IsEmailTest -> IsEmailCat
decodeCat iet = go & override
  where
    go = case ietCat iet of
      "ISEMAIL_ERR" -> CatError
      "ISEMAIL_DEPREC" -> CatDeprec
      _ -> CatOkay
    override cat
      -- We enforce character limits so these will fail.
      | ietId iet == 26 = CatError
      | ietId iet == 28 = CatError
      | ietId iet == 41 = CatError
      -- This is obviously wrong if you put even the slightest amount
      -- of thought into it.
      | ietId iet == 43 = CatError
      -- This one is a bit silly.  Even if that form is deprecated
      -- it would then pass under the general @domain-literal@ rule
      -- in RFC 5322 §3.4.1.
      | ietId iet == 71 = CatOkay
      -- RFC 5322 (and its errata) clearly allows whitespace before
      -- and after the at sign.
      | ietId iet == 85 = CatOkay
      -- It's not worth supporting the very obsolete syntax of having
      -- multiple empty lines before an email address.
      | ietId iet == 89 = CatError
      | ietId iet == 149 = CatError
      -- These should be marked at deprecated.
      | ietId iet == 115 = CatDeprec
      | ietId iet == 116 = CatDeprec
      | ietId iet == 117 = CatDeprec
      -- Obsolete and won't pass validation because they only contain
      -- obsolete characters!
      | ietId iet == 124 = CatError
      | ietId iet == 125 = CatError
      | ietId iet == 134 = CatError
      | ietId iet == 138 = CatError
      | ietId iet == 139 = CatError
      -- Nothing to override.
      | otherwise = cat

rfc5322Examples :: [(Text, Text)]
rfc5322Examples =
  [ ("John Doe <jdoe@machine.example>", "jdoe@machine.example"),
    ("Mary Smith <mary@example.net>", "mary@example.net"),
    ("Michael Jones <mjones@machine.example>", "mjones@machine.example"),
    ("\"Joe Q. Public\" <john.q.public@example.com>", "john.q.public@example.com"),
    ("Mary Smith <mary@x.test>", "mary@x.test"),
    ("jdoe@example.org", "jdoe@example.org"),
    ("Who? <one@y.test>", "one@y.test"),
    ("<boss@nil.test>", "boss@nil.test"),
    ("\"Giant; \\\"Big\\\" Box\" <sysservices@example.net>", "sysservices@example.net"),
    ("Pete <pete@silly.example>", "pete@silly.example"),
    ("Ed Jones <c@a.test>", "c@a.test"),
    ("joe@where.test", "joe@where.test"),
    ("John <jdoe@one.test>", "jdoe@one.test"),
    ("\"Mary Smith: Personal Account\" <smith@home.example>", "smith@home.example"),
    ("Jane Brown <j-brown@other.example>", "j-brown@other.example"),
    ("John <jdoe@one.test> (my dear friend)", "jdoe@one.test")
  ]

rfc5322ObsExamples :: [(Text, Text, Text)]
rfc5322ObsExamples =
  [ ( "Pete(A nice \\) chap) <pete(his account)@silly.test(his host)>",
      "pete@silly.test",
      "Pete (A nice \\) chap) <(his account) pete@silly.test (his host)>"
    ),
    ( "Chris Jones <c@(Chris's host.)public.example>",
      "c@public.example",
      "Chris Jones <c@public.example (Chris's host.)>"
    )
  ]

wikipediaIntExamples :: [Text]
wikipediaIntExamples =
  [ "Pelé@example.com",
    "δοκιμή@παράδειγμα.δοκιμή",
    "我買@屋企.香港",
    "二ノ宮@黒川.日本",
    "медведь@с-балалайкой.рф",
    "संपर्क@डाटामेल.भारत"
  ]

-- | Other examples that should pass.
--
-- > (source, simple-format)
miscExamples :: [Text]
miscExamples =
  [ "example+label@example.com",
    "a@b",
    "simple@example.com",
    "very.common@example.com",
    "disposable.style.email.with+symbol@example.com",
    "other.email-with-hyphen@example.com",
    "fully-qualified-domain@example.com",
    "user.name+tag+sorting@example.com",
    "x@example.com",
    "example-indeed@strange-example.com",
    "admin@mailserver1",
    "example@s.example",
    "\"john..doe\"@example.org",
    "mailhost!username@example.org",
    "user%example.com@example.org"
  ]

genLocalPart :: Gen Text
genLocalPart =
  Gen.filter okay (Gen.text (Range.linear 1 30) unicode)
  where
    okay :: Text -> Bool
    okay t =
      Text.all allowedChar t
        && not (Text.isPrefixOf "." t)
        && not (Text.isSuffixOf "." t)

genDomain :: Gen Text
genDomain =
  Gen.filter okay (Gen.text (Range.linear 1 50) unicode)
  where
    okay :: Text -> Bool
    okay t =
      Text.all allowedChar t
        && not (Text.isPrefixOf "-" t)
        && not (Text.isSuffixOf "-" t)

genShortEmail :: Gen Text
genShortEmail = do
  localPart <- genLocalPart
  domain <- genDomain
  pure (localPart <> "@" <> domain)

genLongEmail :: Gen Text
genLongEmail = do
  display <- genLocalPart
  localPart <- genLocalPart
  domain <- genDomain
  pure $
    mconcat
      [ display,
        " <",
        localPart,
        "@",
        domain,
        ">"
      ]

allowedChar :: Char -> Bool
allowedChar c = isAscii c || utf8NonAscii c

-- | Modified 'unicode' generator from Hedgehog.
unicode :: Gen Char
unicode =
  Gen.frequency
    [ (55296, Gen.element "!#$%&'*+-/=?^_`{|}~"),
      (55296, Gen.alphaNum),
      (8190, Gen.filter isPrint (Gen.enum (chr 0xc2) maxBound))
    ]


test/docs/Main.hs
-- |
--
-- Copyright:
--   This file is part of the package addy. It is subject to the license
--   terms in the LICENSE file found in the top-level directory of this
--   distribution and at:
--
--     https://code.devalot.com/open/addy
--
--   No part of this package, including this file, may be copied,
--   modified, propagated, or distributed except according to the terms
--   contained in the LICENSE file.
--
-- License: BSD-2-Clause
module Main
  ( main,
  )
where

import Test.DocTest

main :: IO ()
main = doctest ["-isrc", "src/Addy.hs"]


src/Addy/Internal/Char.hs
-- |
--
-- Copyright:
--   This file is part of the package addy. It is subject to the license
--   terms in the LICENSE file found in the top-level directory of this
--   distribution and at:
--
--     https://code.devalot.com/open/addy
--
--   No part of this package, including this file, may be copied,
--   modified, propagated, or distributed except according to the terms
--   contained in the LICENSE file.
--
-- License: BSD-2-Clause
--
-- Internal functions representing character classes in email
-- addresses.
--
-- Obsolete characters are only supported in
-- 'Addy.Internal.Parser.Lenient' mode and are filtered out after
-- parsing.
module Addy.Internal.Char
  ( utf8NonAscii,
    obsNoWsCtl,
    wsp,
    vchar,
    atext,
    dtext,
    ctext,
    ctextObs,
    qtext,
    qtextObs,
    quotedPair,
    quotedPairObs,
  )
where

import Data.Char

-- | Is a character in the @UTF8-non-ascii@ class from RFC 6532?
--
-- @since 0.1.0.0
utf8NonAscii :: Char -> Bool
utf8NonAscii c = ord c >= 0xc2 && isPrint c

-- | Obsolete control characters.
--
-- > obs-NO-WS-CTL   =   %d1-8 /            ; US-ASCII control
-- >                     %d11 /             ;  characters that do not
-- >                     %d12 /             ;  include the carriage
-- >                     %d14-31 /          ;  return, line feed, and
-- >                     %d127              ;  white space characters
--
-- @since 0.1.0.0
obsNoWsCtl :: Char -> Bool
obsNoWsCtl = ord >>> go
  where
    go n =
      (n >= 1 && n <= 8)
        || n == 11
        || n == 12
        || (n >= 14 && n <= 31)
        || n == 127

-- | Whitepace.
--
-- @since 0.1.0.0
wsp :: Char -> Bool
wsp c = c == ' ' || c == '\t'

-- | RFC 5234: Visible character.
--
-- > VCHAR          =  %x21-7E
-- >                        ; visible (printing) characters
--
--  RFC 6532 §3.2
--
-- > VCHAR   =/  UTF8-non-ascii
--
-- @since 0.1.0.0
vchar :: Char -> Bool
vchar c = vchar' (ord c) || utf8NonAscii c
  where
    vchar' n = n >= 0x21 && n <= 0x7e

-- | RFC 5322 §3.2.3
--
-- > atext =   ALPHA / DIGIT /    ; Printable US-ASCII
-- >           "!" / "#" /        ;  characters not including
-- >           "$" / "%" /        ;  specials.  Used for atoms.
-- >           "&" / "'" /
-- >           "*" / "+" /
-- >           "-" / "/" /
-- >           "=" / "?" /
-- >           "^" / "_" /
-- >           "`" / "{" /
-- >           "|" / "}" /
-- >           "~"
--
--  RFC 6532 §3.2
--
-- > atext =/  UTF8-non-ascii
--
-- @since 0.1.0.0
atext :: Char -> Bool
atext c =
  isAlphaNum c
    || utf8NonAscii c
    || c == '!'
    || c == '#'
    || c == '$'
    || c == '%'
    || c == '&'
    || c == '\''
    || c == '*'
    || c == '+'
    || c == '-'
    || c == '/'
    || c == '='
    || c == '?'
    || c == '^'
    || c == '_'
    || c == '`'
    || c == '{'
    || c == '|'
    || c == '}'
    || c == '~'

-- | RFC 5322 @dtext@.
--
-- > dtext           =   %d33-90 /          ; Printable US-ASCII
-- >                     %d94-126 /         ;  characters not including
-- >                     obs-dtext          ;  "[", "]", or "\"
-- > obs-dtext       =   obs-NO-WS-CTL / quoted-pair
--
--  RFC 6532 §3.2
--
-- > dtext   =/  UTF8-non-ascii
-- @since 0.1.0.0
dtext :: Char -> Bool
dtext c = asciidtext (ord c) || utf8NonAscii c
  where
    asciidtext n =
      (n >= 33 && n <= 90)
        || (n >= 94 && n <= 126)

-- | RFC 5322 @ctext@.
--
-- > ctext           =   %d33-39 /          ; Printable US-ASCII
-- >                     %d42-91 /          ;  characters not including
-- >                     %d93-126 /         ;  "(", ")", or "\"
-- >                     obs-ctext
-- >
-- > obs-ctext       =   obs-NO-WS-CTL
--
-- RFC 6532 §3.2
--
-- > ctext   =/  UTF8-non-ascii
--
-- @since 0.1.0.0
ctext :: Char -> Bool
ctext c = asciictext (ord c) || utf8NonAscii c
  where
    asciictext n =
      (n >= 33 && n <= 39)
        || (n >= 42 && n <= 91)
        || (n >= 93 && n <= 126)

-- | Obsolete @ctext@.
--
-- > obs-ctext = obs-NO-WS-CTL
--
-- @since 0.1.0.0
ctextObs :: Char -> Bool
ctextObs = obsNoWsCtl

-- | Characters that can appear in a quoted string.
--
-- RFC 5322 §3.2.4:
--
-- > qtext =   %d33 /             ; Printable US-ASCII
-- >           %d35-91 /          ;  characters not including
-- >           %d93-126 /         ;  "\" or the quote character
-- >           obs-qtext
--
-- RFC 6532 §3.2:
--
-- > qtext =/ UTF8-non-ascii
-- @since 0.1.0.0
qtext :: Char -> Bool
qtext c = asciiqtext (ord c) || utf8NonAscii c
  where
    asciiqtext n =
      n == 33
        || (n >= 35 && n <= 91)
        || (n >= 93 && n <= 126)

-- | Obsolete @qtext@.
--
-- > obs-qtext = obs-NO-WS-CTL
--
-- @since 0.1.0.0
qtextObs :: Char -> Bool
qtextObs = obsNoWsCtl

-- | Characters that can follow a backslash.
--
-- > quoted-pair     =   ("\" (VCHAR / WSP)) / obs-qp
--
-- @since 0.1.0.0
quotedPair :: Char -> Bool
quotedPair c = vchar c || wsp c

-- | Obsolete characters that can be escaped with a backslash.
--
-- > obs-qp          =   "\" (%d0 / obs-NO-WS-CTL / LF / CR)
--
-- @since 0.1.0.0
quotedPairObs :: Char -> Bool
quotedPairObs c =
  obsNoWsCtl c
    || c == '\r'
    || c == '\n'
    || c == '\0'


src/Addy/Internal/Render.hs
{-# OPTIONS_GHC -fno-warn-orphans #-}

-- |
--
-- Copyright:
--   This file is part of the package addy. It is subject to the license
--   terms in the LICENSE file found in the top-level directory of this
--   distribution and at:
--
--     https://code.devalot.com/open/addy
--
--   No part of this package, including this file, may be copied,
--   modified, propagated, or distributed except according to the terms
--   contained in the LICENSE file.
--
-- License: BSD-2-Clause
--
-- Internal functions to render an 'EmailAddr' to a 'TB.Builder'.
module Addy.Internal.Render
  ( Mode (..),
    render,
    renderToText,
    renderAddrSpec,
    renderDisplayName,
    renderComments,
  )
where

import Addy.Internal.Char
import Addy.Internal.Types
import qualified Data.Text as Text
import qualified Data.Text.Lazy.Builder as TB
import qualified Net.IP as IP
import Text.Show (Show (..), showParen, showString)
import Prelude hiding (show)

-- | Render mode.
--
-- @since 0.1.0.0
data Mode
  = -- | Render the entire email address, including the optional
    -- display name and comments.
    Full
  | -- | Only render the simplest form of the email address.  Only the
    -- 'LocalPart' and 'Domain' are rendered in this mode.
    Short

-- | Render an email address.
--
-- @since 0.1.0.0
render :: Mode -> EmailAddr -> TB.Builder
render = \case
  Short ->
    renderAddrSpec Short
  Full -> \addr@EmailAddr {..} ->
    case _displayName of
      Nothing ->
        renderAddrSpec Full addr
          <> renderComments Full AfterAddress _comments
      Just name ->
        mconcat
          [ renderComments Full BeforeDisplayName _comments,
            renderDisplayName name <> TB.singleton ' ',
            renderComments Full AfterDisplayName _comments,
            TB.singleton '<' <> renderAddrSpec Full addr <> TB.singleton '>',
            renderComments Full AfterAddress _comments
          ]

-- | Render an email address in @addr-spec@ format.
--
-- @since 0.1.0.0
renderAddrSpec :: Mode -> EmailAddr -> TB.Builder
renderAddrSpec mode EmailAddr {..} =
  mconcat
    [ renderComments mode BeforeLocalPart _comments,
      lp _localPart <> TB.singleton '@' <> dn _domain,
      renderComments mode AfterDomain _comments
    ]
  where
    lp :: LocalPart -> TB.Builder
    lp (LP t)
      | mustQuoteLocalPart t = wrap '"' '"' t
      | otherwise = TB.fromText t
    dn :: Domain -> TB.Builder
    dn = \case
      Domain (DN t) ->
        TB.fromText t
      DomainLiteral lit ->
        wrap '[' ']' $ case lit of
          IpAddressLiteral ip ->
            if IP.isIPv6 ip
              then "IPv6:" <> IP.encode ip
              else IP.encode ip
          TaggedAddressLiteral (AT tag) (Lit body) ->
            tag <> ":" <> body
          AddressLiteral (Lit body) ->
            body

-- | Render a display name.
--
-- @since 0.1.0.0
renderDisplayName :: DisplayName -> TB.Builder
renderDisplayName (DP t)
  | Text.all (\c -> atext c || wsp c) t =
    TB.fromText t
  | otherwise =
    wrap '"' '"' t

-- | Render comments that have the given 'CommentLoc'.  The comment
-- location is also used to decide where to introduce white space.
--
-- @since 0.1.0.0
renderComments :: Mode -> CommentLoc -> [Comment] -> TB.Builder
renderComments Short _ _ = mempty
renderComments Full loc cs =
  case go (== loc) cs of
    Nothing -> mempty
    Just tb -> case loc of
      BeforeDisplayName -> TB.singleton ' ' <> tb
      AfterDisplayName -> tb <> TB.singleton ' '
      BeforeLocalPart -> tb <> TB.singleton ' '
      AfterDomain -> TB.singleton ' ' <> tb
      AfterAddress -> TB.singleton ' ' <> tb
  where
    go :: (CommentLoc -> Bool) -> [Comment] -> Maybe TB.Builder
    go f cs =
      filter (\(Comment loc (CC t)) -> f loc && not (Text.null t)) cs
        & map (\(Comment _ (CC t)) -> t)
        & Text.intercalate " "
        & \t ->
          if Text.null t
            then Nothing
            else Just $ wrap '(' ')' t

-- | Render the given address as text.
--
-- @since 0.1.0.0
renderToText :: Mode -> EmailAddr -> Text
renderToText m =
  render m
    >>> TB.toLazyText
    >>> toStrict

-- | Wrap and quote some text.
--
-- @since 0.1.0.0
wrap :: Char -> Char -> Text -> TB.Builder
wrap lh rh t =
  mconcat
    [ TB.singleton lh,
      Text.foldl' escape mempty t,
      TB.singleton rh
    ]
  where
    escape :: TB.Builder -> Char -> TB.Builder
    escape tb c
      | c == lh || c == rh || c == '\\' =
        tb <> TB.singleton '\\' <> TB.singleton c
      | otherwise =
        tb <> TB.singleton c

-- | 'True' if the give text, when used as the @local-part@ of an
-- email address must be wrapped in quotation marks.
--
-- @since 0.1.0.0
mustQuoteLocalPart :: Text -> Bool
mustQuoteLocalPart name =
  Text.any
    ( \c ->
        c == '"'
          || c == '\\'
          || c == ')'
          || c == '('
          || c == '@'
          || wsp c
    )
    name
    || Text.isPrefixOf "." name
    || Text.isSuffixOf "." name
    || Text.isInfixOf ".." name
    || Text.null name -- Yes, this is totally legit.

-- Orphan instance that renders the email address.
instance Show EmailAddr where
  showsPrec d addr =
    showParen (d > 10) $
      showString "EmailAddr " . showsPrec d (render Full addr)


src/Addy.hs
-- |
--
-- Copyright:
--   This file is part of the package addy. It is subject to the license
--   terms in the LICENSE file found in the top-level directory of this
--   distribution and at:
--
--     https://code.devalot.com/open/addy
--
--   No part of this package, including this file, may be copied,
--   modified, propagated, or distributed except according to the terms
--   contained in the LICENSE file.
--
-- License: BSD-2-Clause
--
-- Email addressed are complicated, really complicated.  This library
-- supports all standardized forms of email addresses, including those
-- with UTF-8 encoded Unicode characters.  The standards used by this
-- library include:
--
--   * RFC 1123: Requirements for Internet Hosts -- Application and Support
--   * RFC 2181: Clarifications to the DNS Specification
--   * RFC 3696: Application Techniques for Checking and Transformation of Names
--   * RFC 5321: Simple Mail Transfer Protocol
--   * RFC 5322: Internet Message Format
--   * RFC 6531: SMTP Extension for Internationalized Email
--   * RFC 6532: Internationalized Email Headers
module Addy
  ( -- * How to use this library
    -- $use

    -- * Decoding and encoding
    decode,
    decodeLenient,
    encode,
    encodeFull,

    -- * Email addresses
    EmailAddr,
    emailAddr,
    emailAddrLit,
    displayName,
    localPart,
    domain,
    comments,

    -- * Display name
    DisplayName,
    _DisplayName,
    validateDisplayName,

    -- * Local part
    LocalPart,
    _LocalPart,
    validateLocalPart,

    -- * Domain part
    Domain (..),
    _Domain,
    _DomainLiteral,
    DomainName,
    _DomainName,
    validateDomainName,

    -- * Host names
    HostName,
    _HostNames,
    _HostName,
    validateHostName,

    -- * Address literals
    AddressLiteral (..),
    _IpAddressLiteral,
    _TaggedAddressLiteral,
    _AddressLiteral,
    AddressTag,
    _AddressTag,
    validateAddressTag,
    Literal,
    _Literal,
    validateLiteral,

    -- * Comments
    Comment (..),
    _Comment,
    commentLoc,
    commentContent,
    CommentLoc (..),
    CommentContent,
    _CommentContent,
    validateCommentContent,
  )
where

import Addy.Internal.Parser as P
import Addy.Internal.Render as R
import Addy.Internal.Types
import Addy.Internal.Validation
import Control.Lens (Prism', prism')
import Validation.Combinators (successToMaybe)

-- | Decode an email address.
--
-- @since 0.1.0.0
decode :: Text -> Either (NonEmpty Error) EmailAddr
decode = P.parseWithMode P.Strict

-- | Decode an email address, allowing obsolete characters.  The
-- obsolete characters are parsed but not included in the output.
--
-- This is useful for exacting email addresses from mail messages but
-- should not be used to validate user input.
--
-- >>> Addy.decode "my . email . addy@(WTF)example.com"
-- Left (ParserFailedError "local part > quoted content > '\"': Failed reading: satisfy" :| [])
--
-- >>> Addy.decodeLenient "my . email . addy@(WTF)example.com"
-- Right (EmailAddr "my.email.addy@example.com (WTF)")
--
-- @since 0.1.0.0
decodeLenient :: Text -> Either (NonEmpty Error) EmailAddr
decodeLenient = P.parseWithMode P.Lenient

-- | Encode an email address as text.  This function produces the
-- short form of an email address.  That is, just the 'LocalPart' and
-- the 'Domain' separated by @\@@.
--
-- @since 0.1.0.0
encode :: EmailAddr -> Text
encode = R.renderToText R.Short

-- | Encode a complete email address to text, including the optional
-- display name and any comments.
--
-- @since 0.1.0.0
encodeFull :: EmailAddr -> Text
encodeFull = R.renderToText R.Full

-- | Build an 'EmailAddr' from a 'LocalPart' and 'DomainName'.
--
-- @since 0.1.0.0
emailAddr :: LocalPart -> DomainName -> EmailAddr
emailAddr l d = EmailAddr Nothing l (Domain d) []

-- | Build an 'EmailAddr' from a 'LocalPart' and an 'AddressLiteral'.
--
-- @since 0.1.0.0
emailAddrLit :: LocalPart -> AddressLiteral -> EmailAddr
emailAddrLit l d = EmailAddr Nothing l (DomainLiteral d) []

-- | Prism for working with display names.
--
-- > import Control.Lens ((^?), review)
--
-- To convert text into a 'DisplayName' with content validation:
--
-- >>> "Some Text" ^? _DisplayName
-- Just (DisplayName "Some Text")
-- >>> "Some\nText" ^? _DisplayName
-- Nothing -- Validation failed.
--
-- To access the text content of a 'DisplayName':
--
-- >>> review _DisplayName someDisplayName
-- "Some Text"
--
-- Uses 'validateDisplayName' to perform validation.
--
-- @since 0.1.0.0
_DisplayName :: Prism' Text DisplayName
_DisplayName =
  prism'
    displayNameText
    (validateDisplayName >>> successToMaybe)

-- | Prism for working with the local part of an email address.
--
-- > import Control.Lens ((^?), review)
--
-- To convert text to a 'LocalPart' with content validation:
--
-- >>> "cockroach+mouse" ^? _LocalPart
-- Just (LocalPart "cockroach+mouse")
-- >>> "cockroach\nmouse" ^? _LocalPart
-- Nothing -- Validation failed.
--
-- To access the text content of a 'LocalPart':
--
-- >>> review _LocalPart someLocalPart
-- "cockamouse"
--
-- Uses 'validateLocalPart' to perform validation.
--
-- @since 0.1.0.0
_LocalPart :: Prism' Text LocalPart
_LocalPart =
  prism'
    localPartText
    (validateLocalPart >>> successToMaybe)

-- | Prism for working with domain names.
--
-- > import Control.Lens ((^?), review)
--
-- To convert text to a 'DomainName' with validation:
--
-- >>> "gmail.com" ^? _DomainName
-- Just (DomainName "gmail.com")
-- >>> "too.many.dots." ^? _DomainName
-- Nothing
--
-- To access the text content of a 'DomainName':
--
-- >>> review _DomainName someDomainName
-- "gmail.com"
--
-- Uses 'validateDomainName' to perform validation.
--
-- @since 0.1.0.0
_DomainName :: Prism' Text DomainName
_DomainName =
  prism'
    domainNameText
    (validateDomainName >>> successToMaybe)

-- | Prism for working with host names (DNS labels).
--
-- > import Control.Lens ((^?), review)
--
-- To convert text to a host name with validation:
--
-- >>> "com" ^? _HostName
-- Just (HostName "com")
-- >>> "com." ^? _HostName
-- Nothing -- Validation failed.
--
-- To access the text content of a 'HostName':
--
-- >>> review _HostName someHostName
-- "com"
--
-- Uses 'validateHostName' to perform validation.
--
-- @since 0.1.0.0
_HostName :: Prism' Text HostName
_HostName =
  prism'
    hostNameText
    (validateHostName >>> successToMaybe)

-- | Prism for working with the 'AddressTag' for an 'AddressLiteral'.
--
-- > import Control.Lens ((^?), review)
--
-- To convert text to an address tag with validation:
--
-- >>> "IPv6" ^? _AddressTag
-- Just (AddressTag "IPv6")
-- >>> "[IPv6]" ^? _AddressTag
-- Nothing -- Validation failed.
--
-- To access the text content of an 'AddressTag':
--
-- >>> review _AddressTag someTag
-- "tag"
--
-- Uses 'validateAddressTag' to perform validation.
--
-- @since 0.1.0.0
_AddressTag :: Prism' Text AddressTag
_AddressTag =
  prism'
    addressTagText
    (validateAddressTag >>> successToMaybe)

-- | Prism for working with the literal text of an address literal.
--
-- > import Control.Lens ((^?), review)
--
-- To convert text to an address literal with validation:
--
-- >>> "127.0.0.1" ^? _Literal
-- Just (Literal "127.0.0.1")
-- >>> "[]" ^? _Literal
-- Nothing -- Validation failed.
--
-- To access the text content of a 'Literal':
--
-- >>> review _Literal someLiteral
-- "127.0.0.1"
--
-- Uses 'validateLiteral' to perform validation.
--
-- @since 0.1.0.0
_Literal :: Prism' Text Literal
_Literal =
  prism'
    literalText
    (validateLiteral >>> successToMaybe)

-- | Prism for working with the text content of a comment.
--
-- > import Control.Lens ((^?), review)
--
-- To convert text to a comment with validation:
--
-- >>> "best email ever" ^? _CommentContent
-- Just (CommentContent "best email ever")
-- >>> "\n" ^? _CommentContent
-- Nothing
--
-- To access the text content of the comment:
--
-- >>> review _CommentContent someComment
-- "super"
--
-- Uses 'validateCommentContent' to perform validation.
--
-- @since 0.1.0.0
_CommentContent ::
  Prism' Text CommentContent
_CommentContent =
  prism'
    commentContentText
    (validateCommentContent >>> successToMaybe)

-- $use
--
-- == Importing
--
-- This library is designed to be imported qualified:
--
-- > import qualified Addy
--
-- == Decoding addresses
--
-- To decode (parse) an email address from text:
--
-- >>> Addy.decode "example@example.com"
-- Right (EmailAddr "example@example.com")
--
-- >>> Addy.decode "我買@屋企.香港"
-- Right (EmailAddr "\25105\36023@\23627\20225.\39321\28207")
--
-- >>> Addy.decode "Mary Smith <mary@example.net> (hi there!)"
-- Right (EmailAddr "Mary Smith <mary@example.net> (hi there!)")
--
-- >>> Addy.decode "example@[127.0.0.1]"
-- Right (EmailAddr "example@[127.0.0.1]")
--
-- After decoding, an address is automatically normalized by performing
-- NFC normalization and down-casing the domain name:
--
-- >>> Addy.decode "My.Email.Addy@ExAmPlE.COM"
-- Right (EmailAddr "My.Email.Addy@example.com")
--
-- == Encoding addresses
--
-- Turning an email address back to text is just as easy:
--
-- >>> Addy.encode address
-- "example@example.com"
--
-- If an address has an optional display name or comments you can
-- render those with the 'encodeFull' function.
--
-- >>> :{
--   Addy.decode "Mary Smith <mary@example.net> (hi there!)"
--     & second Addy.encodeFull
-- :}
-- Right "Mary Smith <mary@example.net> (hi there!)"
--
-- == Creating addresses
--
-- In order to prevent invalid email addresses from being created this
-- library uses @newtype@ wrappers and does not export the data
-- constructors.  Therefore you'll need to use the smart constructor
-- approach using the 'emailAddr' function.
--
-- If you want to work with the validation functions directly we
-- recommend 'Applicative' syntax:
--
-- >>> :{
--  Addy.emailAddr
--    <$> Addy.validateLocalPart "pjones"
--    <*> Addy.validateDomainName "devalot.com"
-- :}
-- Success (EmailAddr "pjones@devalot.com")
--
-- Prisms for the @newtype@ wrappers are provided if you want to use optics:
--
-- >>> :{
--  Addy.emailAddr
--    <$> "pjones" ^? Addy._LocalPart
--    <*> "devalot.com" ^? Addy._DomainName
-- :}
-- Just (EmailAddr "pjones@devalot.com")
--
-- == Optics
--
-- Lens and prisms are provided to make working with email addresses easier:
--
-- > import qualified Addy
-- > import Control.Lens
--
-- >>> Addy.decode "example@example.com" ^? _Right . Addy.domain
-- Just (Domain (DomainName "example.com"))
--
-- >>> Addy.decode "example@example.com"
--   ^? _Right . Addy.domain . Addy._Domain . Addy._HostNames
-- Just [HostName "example",HostName "com"]
--
-- >>> Addy.decode "example@example.com"
--   ^.. _Right . Addy.domain . Addy._Domain
--   . Addy._HostNames . traversed . re _HostName
-- ["example","com"]
--
-- == A word about address literals
--
-- Believe it or not, this is a completely valid email address:
--
-- >>> Addy.decode "example@[what's my domain name?]"
--   ^? _Right . Addy.domain
-- Just (DomainLiteral (AddressLiteral (Literal "what's my domain name?")))
--
-- If you're working with email messages it might be useful to capture
-- these address literals, especially if you know how to interpret
-- them.  However, if you're validating user input you probably don't
-- want to allow these.
--
-- >>> Addy.decode "e@[127.0.0.1]" ^? _Right
--   >>= failover (Addy.domain . Addy._Domain) id
-- Nothing
--
-- >>> Addy.decode "example@example.com" ^? _Right
--   >>= failover (Addy.domain . Addy._Domain) id
-- Just (EmailAddr "example@example.com")
